### PR TITLE
split the trace tags out from the core context ID in marshal/unmarshal

### PR DESCRIPTION
**Pull request by bensigelman**
(Per discussion in #7)

### CODE REVIEW

File: opentracing/tracecontext.go [hunk 1]
```diff
@@ -63,10 +90,31 @@ type TraceContextMarshaler interface {
...
 	// TraceContextMarshaler.MarshalTraceContextStringMap()) into a TraceContext.
-	UnmarshalTraceContextStringMap(marshaled map[string]string) (TraceContext, error)
+	//
+	// The first parameter contains the marshaler's serialization of the core
+	// identifying information in a TraceContext instance.
+	//
+	// The first parameter contains the marshaler's serialization of the trace
+	// tags (per `SetTraceTag` and `TraceTag`) attached to a TraceContext
+	// instance.
+	UnmarshalTraceContextStringMap(
```
Comment (yurishkuro) [inline opentracing/tracecontext.go]: Perhaps add a comment that it's ok to pass the same map as a value for both parameters, e.g. http headers map. The method only takes those keys from the map that it deems relevant.

File: opentracing/tracecontext.go [hunk 2]
```diff
@@ -63,10 +90,31 @@ type TraceContextMarshaler interface {
...
+	) (TraceContext, error)
+
 	// Converts the marshaled string:string map (see
 	// TraceContextMarshaler.MarshalTraceContextStringMap()) into a TraceContext.
-	UnmarshalTraceContextStringMap(marshaled map[string]string) (TraceContext, error)
+	//
+	// The first parameter contains the marshaler's serialization of the core
+	// identifying information in a TraceContext instance.
+	//
+	// The first parameter contains the marshaler's serialization of the trace
```
Comment (yurishkuro) [inline opentracing/tracecontext.go]: s/first/second/

**Comment by yurishkuro** [comment]
looks great.

**Comment by bensigelman** [comment]
done and done -- thanks.

File: examples/dapperish/dapper.go [hunk 3]
```diff
@@ -63,20 +62,20 @@ func (d *DapperishTraceContext) NewChild() (opentracing.TraceContext, opentracin
 }
 
 // SetTraceTag complies with the opentracing.TraceContext interface.
-func (d *DapperishTraceContext) SetTraceTag(key, val string) opentracing.TraceContext {
+func (d *DapperishTraceContext) SetTraceTag(caseInsensitiveKey, val string) opentracing.TraceContext {
```
Comment (adriancole) [inline examples/dapperish/dapper.go]: I'd prefer we elaborate on this somewhere, even if this is a separate issue. For example [http/2 treats transmission of mixed case keys an error](https://http2.github.io/http2-spec/#HttpHeaders), even if comparison (at app layer) is permitted to be insensitive.


> Just as in HTTP/1.x, header field names are strings of ASCII characters that are compared in a case-insensitive fashion. However, header field names MUST be converted to lowercase prior to their encoding in HTTP/2. A request or response containing uppercase header field names MUST be treated as malformed (Section 8.1.2.6).

File: opentracing/tracecontext.go [hunk 4]
```diff
@@ -39,12 +39,16 @@ type TraceContext interface {
 	// TraceContext, and that can add up to a lot of network and cpu
 	// overhead.
 	//
+	// IMPORTANT NOTE #3: Trace tags are case-insensitive: implementations may
```
Comment (adriancole) [inline opentracing/tracecontext.go]: this is where my comment meant to be wrt lc transmission. Impact is that case preservation of keys is unsupported.

File: opentracing/utils.go [hunk 5]
```diff
@@ -7,9 +7,12 @@ import (
 )
 
 const (
-	// OpenTracingContextHTTPHeaderPrefix precedes all opentracing-related HTTP
-	// headers.
-	OpenTracingContextHTTPHeaderPrefix = "Opentracing-Context-"
+	// OpenTracingContextIDHTTPHeaderPrefix precedes the opentracing-related
+	// ContextID HTTP headers.
+	OpenTracingContextIDHTTPHeaderPrefix = "Open-Tracing-Context-Id-"
```
Comment (adriancole) [inline opentracing/utils.go]: let's downcase these, knowing that http/2 will

**Comment by adriancole** [comment]
LGTM

**Pull request merged**